File: plugins/lylacore-native/benchmarks/nodejs/benchmark_soul_crypto.js
// Authors: Joysusy & Violet Klaudia 💖
/**
 * Soul Crypto Performance Benchmarks
 *
 * Comprehensive benchmarking suite for soul-crypto Node.js bindings.
 * Compares performance against Rust baseline and measures FFI overhead.
 */

const { performance } = require('perf_hooks');

// Import the native addon
let soulCrypto;
try {
    soulCrypto = require('../../crates/napi-bindings');
} catch (error) {
    console.error('Error: lylacore native addon not found. Build with: npm run build');
    process.exit(1);
}

/**
 * Run a benchmark with adaptive iterations
 */
function benchmark(name, fn, minIterations = 10, targetTime = 1000) {
    // Warm-up
    for (let i = 0; i < 5; i++) {
        fn();
    }

    // Determine iteration count
    const start = performance.now();
    fn();
    const singleTime = performance.now() - start;

    const iterations = Math.max(minIterations, Math.floor(targetTime / singleTime));

    // Actual benchmark
    const benchStart = performance.now();
    for (let i = 0; i < iterations; i++) {
        fn();
    }
    const benchEnd = performance.now();

    const avgTime = (benchEnd - benchStart) / iterations;

    // Format output
    let timeStr;
    if (avgTime < 0.001) {
        timeStr = `${(avgTime * 1e6).toFixed(2)} ns`;
    } else if (avgTime < 1) {
        timeStr = `${(avgTime * 1e3).toFixed(2)} µs`;
    } else {
        timeStr = `${avgTime.toFixed(2)} ms`;
    }

    console.log(`${name.padEnd(30)} | ${timeStr.padStart(12)}`);
}

/**
 * Benchmark key derivation algorithms
 */
function benchmarkKeyDerivation() {
    console.log('\n=== Key Derivation Benchmarks ===\n');

    const passphrase = 'benchmark_passphrase_for_testing';
    const salt = soulCrypto.generateSalt();

    const algorithms = [
        ['Argon2id (sync)', 'argon2id', true],
        ['Argon2id (async)', 'argon2id', false],
        ['PBKDF2 (sync)', 'pbkdf2', true],
        ['Scrypt (sync)', 'scrypt', true],
    ];

    for (const [name, algo, useSync] of algorithms) {
        if (useSync) {
            benchmark(name, () => {
                soulCrypto.deriveKeySync(passphrase, salt, { algorithm: algo });
            }, 10);
        } else {
            // For async, we need to handle promises
            console.log(`${name.padEnd(30)} | (async - run separately)`);
        }
    }
}

/**
 * Benchmark encryption operations
 */
function benchmarkEncryption() {
    console.log('\n=== Encryption Benchmarks ===\n');

    const salt = soulCrypto.generateSalt();
    const key = soulCrypto.deriveKeySync('test', salt, { algorithm: 'pbkdf2' });

    const sizes = [
        [64, '64B'],
        [1024, '1KB'],
        [64 * 1024, '64KB'],
        [1024 * 1024, '1MB'],
    ];

    for (const [size, label] of sizes) {
        const plaintext = Buffer.alloc(size, 0x61);

        benchmark(`Encrypt ${label}`, () => {
            soulCrypto.encrypt(key, plaintext);
        }, size <= 1024 ? 1000 : 100);
    }
}

/**
 * Benchmark decryption operations
 */
function benchmarkDecryption() {
    console.log('\n=== Decryption Benchmarks ===\n');

    const salt = soulCrypto.generateSalt();
    const key = soulCrypto.deriveKeySync('test', salt, { algorithm: 'pbkdf2' });

    const sizes = [
        [64, '64B'],
        [1024, '1KB'],
        [64 * 1024, '64KB'],
        [1024 * 1024, '1MB'],
    ];

    for (const [size, label] of sizes) {
        const plaintext = Buffer.alloc(size, 0x61);
        const { nonce, ciphertext, authTag } = soulCrypto.encrypt(key, plaintext);

        benchmark(`Decrypt ${label}`, () => {
            soulCrypto.decrypt(key, nonce, ciphertext, authTag);
        }, size <= 1024 ? 1000 : 100);
    }
}

/**
 * Benchmark full workflow
 */
function benchmarkFullWorkflow() {
    console.log('\n=== Full Workflow Benchmark ===\n');

    const passphrase = 'workflow_benchmark_passphrase';
    const plaintext = Buffer.from('Sensitive data for benchmarking');

    benchmark('Derive → Encrypt → Decrypt', () => {
        const salt = soulCrypto.generateSalt();
        const key = soulCrypto.deriveKeySync(passphrase, salt, { algorithm: 'argon2id' });
        const { nonce, ciphertext, authTag } = soulCrypto.encrypt(key, plaintext);
        soulCrypto.decrypt(key, nonce, ciphertext, authTag);
    }, 10);
}

/**
 * Benchmark async key derivation separately
 */
async function benchmarkAsyncKeyDerivation() {
    console.log('\n=== Async Key Derivation Benchmark ===\n');

    const passphrase = 'benchmark_passphrase_for_testing';
    const salt = soulCrypto.generateSalt();

    // Warm-up
    for (let i = 0; i < 3; i++) {
        await soulCrypto.deriveKey(passphrase, salt, { algorithm: 'argon2id' });
    }

    // Benchmark
    const iterations = 10;
    const start = performance.now();

    for (let i = 0; i < iterations; i++) {
        await soulCrypto.deriveKey(passphrase, salt, { algorithm: 'argon2id' });
    }

    const end = performance.now();
    const avgTime = (end - start) / iterations;

    console.log(`${'Argon2id (async)'.padEnd(30)} | ${avgTime.toFixed(2).padStart(12)} ms`);
}

/**
 * Main benchmark runner
 */
async function main() {
    console.log('='.repeat(60));
    console.log('Soul Crypto Node.js Bindings - Performance Benchmarks');
    console.log('='.repeat(60));

    benchmarkKeyDerivation();
    benchmarkEncryption();
    benchmarkDecryption();
    benchmarkFullWorkflow();

    await benchmarkAsyncKeyDerivation();

    console.log('\n' + '='.repeat(60));
    console.log('Benchmarks Complete');
    console.log('='.repeat(60));
}

// Run benchmarks
main().catch(console.error);

File: plugins/lylacore-native/benchmarks/python/benchmark_soul_crypto.py
#!/usr/bin/env python3
# Authors: Joysusy & Violet Klaudia 💖
"""
Soul Crypto Performance Benchmarks

Comprehensive benchmarking suite for soul-crypto Python bindings.
Compares performance against Rust baseline and measures FFI overhead.
"""

import sys
import timeit
from pathlib import Path

# Add parent directory to path to import lylacore
sys.path.insert(0, str(Path(__file__).parent.parent.parent / "crates" / "pyo3-bindings"))

try:
    from lylacore import soul_crypto
except ImportError:
    print("Error: lylacore module not found. Build with: maturin develop")
    sys.exit(1)


def benchmark_key_derivation():
    """Benchmark key derivation algorithms."""
    print("\n=== Key Derivation Benchmarks ===\n")

    passphrase = "benchmark_passphrase_for_testing"
    salt = soul_crypto.generate_salt()

    algorithms = [
        ("Argon2id", "argon2id"),
        ("PBKDF2", "pbkdf2"),
        ("Scrypt", "scrypt"),
    ]

    for name, algo in algorithms:
        def bench():
            soul_crypto.derive_key(passphrase, salt, algorithm=algo)

        # Run with adaptive iterations (fewer for slow operations)
        iterations = 10 if algo == "argon2id" else 100
        time_taken = timeit.timeit(bench, number=iterations) / iterations

        print(f"{name:12} | {time_taken * 1000:8.2f} ms")


def benchmark_encryption():
    """Benchmark encryption operations."""
    print("\n=== Encryption Benchmarks ===\n")

    key = soul_crypto.derive_key("test", soul_crypto.generate_salt(), algorithm="pbkdf2")

    sizes = [
        (64, "64B"),
        (1024, "1KB"),
        (64 * 1024, "64KB"),
        (1024 * 1024, "1MB"),
    ]

    for size, label in sizes:
        plaintext = b"a" * size

        def bench():
            soul_crypto.encrypt(key, plaintext)

        # Adaptive iterations based on size
        if size <= 1024:
            iterations = 10000
        elif size <= 64 * 1024:
            iterations = 1000
        else:
            iterations = 100

        time_taken = timeit.timeit(bench, number=iterations) / iterations

        if time_taken < 1e-6:
            print(f"Encrypt {label:6} | {time_taken * 1e9:8.2f} ns")
        elif time_taken < 1e-3:
            print(f"Encrypt {label:6} | {time_taken * 1e6:8.2f} µs")
        else:
            print(f"Encrypt {label:6} | {time_taken * 1e3:8.2f} ms")


def benchmark_decryption():
    """Benchmark decryption operations."""
    print("\n=== Decryption Benchmarks ===\n")

    key = soul_crypto.derive_key("test", soul_crypto.generate_salt(), algorithm="pbkdf2")

    sizes = [
        (64, "64B"),
        (1024, "1KB"),
        (64 * 1024, "64KB"),
        (1024 * 1024, "1MB"),
    ]

    for size, label in sizes:
        plaintext = b"a" * size
        nonce, ciphertext, auth_tag = soul_crypto.encrypt(key, plaintext)

        def bench():
            soul_crypto.decrypt(key, nonce, ciphertext, auth_tag)

        # Adaptive iterations based on size
        if size <= 1024:
            iterations = 10000
        elif size <= 64 * 1024:
            iterations = 1000
        else:
            iterations = 100

        time_taken = timeit.timeit(bench, number=iterations) / iterations

        if time_taken < 1e-6:
            print(f"Decrypt {label:6} | {time_taken * 1e9:8.2f} ns")
        elif time_taken < 1e-3:
            print(f"Decrypt {label:6} | {time_taken * 1e6:8.2f} µs")
        else:
            print(f"Decrypt {label:6} | {time_taken * 1e3:8.2f} ms")


def benchmark_full_workflow():
    """Benchmark complete derive → encrypt → decrypt workflow."""
    print("\n=== Full Workflow Benchmark ===\n")

    passphrase = "workflow_benchmark_passphrase"
    plaintext = b"Sensitive data for benchmarking"

    def bench():
        salt = soul_crypto.generate_salt()
        key = soul_crypto.derive_key(passphrase, salt, algorithm="argon2id")
        nonce, ciphertext, auth_tag = soul_crypto.encrypt(key, plaintext)
        soul_crypto.decrypt(key, nonce, ciphertext, auth_tag)

    iterations = 10
    time_taken = timeit.timeit(bench, number=iterations) / iterations

    print(f"Derive → Encrypt → Decrypt | {time_taken * 1000:8.2f} ms")


def main():
    """Run all benchmarks."""
    print("=" * 60)
    print("Soul Crypto Python Bindings - Performance Benchmarks")
    print("=" * 60)

    benchmark_key_derivation()
    benchmark_encryption()
    benchmark_decryption()
    benchmark_full_workflow()

    print("\n" + "=" * 60)
    print("Benchmarks Complete")
    print("=" * 60)


if __name__ == "__main__":
    main()
